google/api_core/retry_streaming.py

```diff
@@ -0,0 +1,183 @@
+        check if it is retryable. If so, create a new active_target and
+        continue iterating. If not, raise the exception.
+        """
+        if not self.predicate(exc):
+            raise exc
+        else:
+            # run on_error callback if provided
+            if self.on_error:
+                self.on_error(exc)
+            try:
+                next_sleep = next(self.sleep_generator)
+            except StopIteration:
+                raise ValueError("Sleep generator stopped yielding sleep values")
+            # if deadline is exceeded, raise RetryError
+            if self.deadline is not None:
+                next_attempt = datetime_helpers.utcnow() + datetime.timedelta(
+                    seconds=next_sleep
+                )
+                if self.deadline < next_attempt:
+                    raise exceptions.RetryError(
+                        f"Deadline of {self.timeout:.1f} seconds exceeded", exc
+                    ) from exc
+            # sleep before retrying
+            _LOGGER.debug(
+                "Retrying due to {}, sleeping {:.1f}s ...".format(exc, next_sleep)
+            )
+            time.sleep(next_sleep)
+            self.active_target = self.target_fn()
+
+    def __next__(self):
+        """
+        Implement the iterator protocol.
+        """
+        try:
+            return next(self.active_target)
+        except Exception as exc:
+            self._handle_exception(exc)
+        # if retryable exception was handled, try again with new active_target
+        return self.__next__()
+
+    def close(self):
+        """
+        Close the active_target if supported. (e.g. target is a generator)
+
+        Raises:
+          - AttributeError if the active_target does not have a close() method
+        """
+        if getattr(self.active_target, "close", None):
+            return self.active_target.close()
+        else:
+            raise AttributeError(
+                "close() not implemented for {}".format(self.active_target)
+            )
+
+    def send(self, value):
+        """
+        Call send on the active_target if supported. (e.g. target is a generator)
+
+        If an exception is raised, a retry may be attempted before returning
+        a result.
+
+        Returns:
+          - the result of calling send() on the active_target
+
+        Raises:
+          - AttributeError if the active_target does not have a send() method
+        """
+        if getattr(self.active_target, "send", None):
+            try:
+                return self.active_target.send(value)
+            except Exception as exc:
+                self._handle_exception(exc)
+            # if exception was retryable, use new target for return value
+            return self.__next__()
+        else:
+            raise AttributeError(
+                "send() not implemented for {}".format(self.active_target)
+            )
+
+    def throw(self, typ, val=None, tb=None):
```

Comment:
Please add type hints for these args. Please document the args in the docstring. Use more descriptive arg names. eg. What is `tb`?

Reply:
`send`, `throw`, `next`, and `close` are all part of the generator spec, so this signature comes from there. 

I'm a bit hesitant to spend too much time documenting the generator spec here. Maybe we should just use *args and **kwargs and pass them instead?